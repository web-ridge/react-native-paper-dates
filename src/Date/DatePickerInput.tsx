import { TextInput } from 'react-native-paper'
import DatePickerModal from './DatePickerModal'
import { useLatest } from '../shared/utils'
import type { DatePickerInputProps } from './DatePickerInput.shared'
import DatePickerInputWithoutModal from './DatePickerInputWithoutModal'
import { Platform, StyleProp, ViewStyle } from 'react-native'
import { forwardRef, useCallback, useState } from 'react'
import { defaultStartYear, defaultEndYear } from './dateUtils'

function DatePickerInput(
  {
    withModal = true,
    calendarIcon = 'calendar',
    disableCalendarIcon = false,
    animationType = Platform.select({ web: 'none', default: 'slide' }),
    presentationStyle = 'overFullScreen',
    ...rest
  }: DatePickerInputProps,
  ref: any
) {
  const [visible, setVisible] = useState(false)

  const onChangeRef = useLatest(rest.onChange)

  const onDismiss = useCallback(() => {
    setVisible(false)
  }, [setVisible])

  const onInnerConfirm = useCallback(
    ({ date }: any) => {
      setVisible(false)
      onChangeRef.current(date)
    },
    [setVisible, onChangeRef]
  )

  return (
    <DatePickerInputWithoutModal
      ref={ref}
      {...rest}
      inputButton={
        withModal ? (
          <TextInput.Icon
            size={rest.iconSize ?? 24}
            icon={calendarIcon}
            color={rest.iconColor ?? undefined}
            forceTextInputFocus={false}
            disabled={rest.disabled || disableCalendarIcon}
            onPress={() => setVisible(true)}
            style={rest.iconStyle as StyleProp<ViewStyle>}
            testID={`${rest.testID || 'date-picker'}-icon-button`}
          />
        ) : null
      }
      // eslint-disable-next-line react/no-unstable-nested-components
      modal={({
        value,
        locale,
        inputMode,
        validRange,
        saveLabel,
        saveLabelDisabled,
        uppercase,
        startYear,
        endYear,
        inputEnabled,
        disableStatusBarPadding,
        startWeekOnMonday,
      }) =>
        withModal ? (
          <DatePickerModal
            date={value}
            mode="single"
            visible={visible}
            onDismiss={onDismiss}
            onConfirm={onInnerConfirm}
            locale={locale}
            dateMode={inputMode}
            validRange={validRange}
            saveLabel={saveLabel}
            saveLabelDisabled={saveLabelDisabled ?? false}
            uppercase={uppercase}
            startYear={startYear ?? defaultStartYear}
            endYear={endYear ?? defaultEndYear}
            inputEnabled={inputEnabled}
            disableStatusBarPadding={disableStatusBarPadding ?? false}
            animationType={animationType}
            presentationStyle={presentationStyle}
            label={rest.label as any}
            startWeekOnMonday={startWeekOnMonday}
            withDateFormatInLabel={rest.withDateFormatInLabel}
            placeholder={rest.placeholder}
          />
        ) : null
      }
    />
  )
}

export default forwardRef(DatePickerInput)
