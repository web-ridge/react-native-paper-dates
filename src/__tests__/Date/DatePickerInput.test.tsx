import { render } from '@testing-library/react-native'
import DatePickerInput from '../../Date/DatePickerInput'

// 7th of August, 2025
const fixedDate = new Date(2025, 7, 7)

it('renders DatePickerInput', () => {
  const { toJSON } = render(
    <DatePickerInput
      locale="en"
      value={fixedDate}
      onChange={() => null}
      inputMode="start"
    />
  )
  expect(toJSON()).toMatchSnapshot()
})
