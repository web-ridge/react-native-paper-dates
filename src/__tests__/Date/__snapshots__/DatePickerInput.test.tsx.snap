// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DatePickerInput 1`] = `
[
  <View
    style={
      {
        "alignItems": "flex-start",
        "flex": 1,
        "flexGrow": 1,
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexGrow": 1,
          "justifyContent": "flex-start",
          "width": "100%",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(231, 224, 236, 1)",
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
            },
            {
              "flexGrow": 1,
              "width": "100%",
            },
          ]
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "bottom": 0,
                "height": 2,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "zIndex": 1,
              },
              {
                "height": 1,
              },
              {
                "backgroundColor": "rgba(73, 69, 79, 1)",
                "transform": [
                  {
                    "scaleY": 0.5,
                  },
                ],
              },
              undefined,
            ]
          }
          testID="text-input-underline"
        />
        <View
          onLayout={[Function]}
          style={
            [
              {
                "flexGrow": 1,
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              {
                "minHeight": 56,
              },
            ]
          }
        >
          <View
            pointerEvents="none"
            style={
              [
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                {
                  "overflow": "hidden",
                },
                {
                  "zIndex": 3,
                },
              ]
            }
          >
            <View
              collapsable={false}
              pointerEvents="none"
              style={
                {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": [
                    {
                      "translateX": 4,
                    },
                  ],
                  "width": 750,
                }
              }
            >
              <View
                style={
                  {
                    "width": 97.33333333333333,
                  }
                }
              >
                <Text
                  collapsable={false}
                  maxFontSizeMultiplier={1.5}
                  numberOfLines={1}
                  onLayout={[Function]}
                  onTextLayout={[Function]}
                  style={
                    {
                      "color": "rgba(103, 80, 164, 1)",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": undefined,
                      "left": 0,
                      "letterSpacing": 0.15,
                      "lineHeight": undefined,
                      "maxWidth": 97.33333333333333,
                      "opacity": 0,
                      "paddingLeft": 16,
                      "paddingRight": 56,
                      "position": "absolute",
                      "textAlign": "left",
                      "top": 30,
                      "transform": [
                        {
                          "translateX": 0,
                        },
                        {
                          "translateY": -12,
                        },
                        {
                          "scale": 0.75,
                        },
                      ],
                      "writingDirection": "ltr",
                    }
                  }
                  testID="text-input-flat-label-active"
                >
                  MM/DD/YYYY
                </Text>
                <Text
                  collapsable={false}
                  maxFontSizeMultiplier={1.5}
                  numberOfLines={1}
                  style={
                    {
                      "color": "rgba(73, 69, 79, 1)",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": undefined,
                      "left": 0,
                      "letterSpacing": 0.15,
                      "lineHeight": undefined,
                      "maxWidth": 97.33333333333333,
                      "opacity": 0,
                      "paddingLeft": 16,
                      "paddingRight": 56,
                      "position": "absolute",
                      "textAlign": "left",
                      "top": 30,
                      "transform": [
                        {
                          "translateX": 0,
                        },
                        {
                          "translateY": -12,
                        },
                        {
                          "scale": 0.75,
                        },
                      ],
                      "writingDirection": "ltr",
                    }
                  }
                  testID="text-input-flat-label-inactive"
                >
                  MM/DD/YYYY
                </Text>
              </View>
            </View>
          </View>
          <TextInput
            cursorColor="rgba(103, 80, 164, 1)"
            editable={true}
            keyboardAppearance="default"
            keyboardType="number-pad"
            maxFontSizeMultiplier={1.5}
            maxLength={10}
            multiline={false}
            onBlur={[Function]}
            onChange={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholderTextColor="transparent"
            selectionColor="rgba(103, 80, 164, 1)"
            style={
              [
                {
                  "flexGrow": 1,
                  "margin": 0,
                },
                {},
                {
                  "paddingBottom": 4,
                  "paddingTop": 24,
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "minWidth": 65,
                  "paddingLeft": 16,
                  "paddingRight": 56,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                undefined,
                {
                  "marginRight": 40,
                  "paddingRight": 16,
                },
                undefined,
              ]
            }
            testID="text-input-flat"
            underlineColorAndroid="transparent"
            value="06/18/2025"
          />
        </View>
        <View
          style={
            [
              {
                "alignItems": "center",
                "height": 24,
                "justifyContent": "center",
                "position": "absolute",
                "width": 24,
              },
              {
                "right": 16,
                "top": 16,
              },
            ]
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "transparent",
                "borderRadius": 20,
                "height": 40,
                "margin": 6,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
                "width": 40,
              }
            }
            testID="date-picker-icon-button-container-outer-layer"
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "transparent",
                  "borderColor": "rgba(121, 116, 126, 1)",
                  "borderRadius": 20,
                  "borderWidth": 0,
                  "elevation": 0,
                  "flex": 1,
                  "overflow": "hidden",
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
              testID="date-picker-icon-button-container"
            >
              <View
                accessibilityComponentType="button"
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityTraits="button"
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                centered={true}
                collapsable={false}
                focusable={true}
                hitSlop={
                  {
                    "bottom": 6,
                    "left": 6,
                    "right": 6,
                    "top": 6,
                  }
                }
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "overflow": "hidden",
                    },
                    [
                      {
                        "alignItems": "center",
                        "flexGrow": 1,
                        "justifyContent": "center",
                      },
                      undefined,
                    ],
                  ]
                }
                testID="date-picker-icon-button"
              >
                <Text
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    [
                      {
                        "backgroundColor": "transparent",
                      },
                      {
                        "color": "rgba(73, 69, 79, 1)",
                        "fontSize": 24,
                      },
                    ]
                  }
                >
                  □
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />,
]
`;
